src/library/scala/collection/immutable/ListSet.scala

```diff
@@ -115,33 +88,34 @@ sealed class ListSet[A] extends AbstractSet[A]
    *  @throws java.util.NoSuchElementException
    *  @return the new iterator
    */
-  def iterator: Iterator[A] = new AbstractIterator[A] {
-    var that: ListSet[A] = self
-    def hasNext = that.nonEmpty
-    def next: A =
-      if (hasNext) {
-        val res = that.head
-        that = that.tail
-        res
+  def iterator: Iterator[A] = {
+    def reverseList = {
+      var curr: ListSet[A] = self
+      var res: List[A] = Nil
+      while (!curr.isEmpty) {
+        res = curr.elem :: res
+        curr = curr.next
       }
-      else Iterator.empty.next()
+      res
+    }
+    reverseList.iterator
   }
 
   /**
    *  @throws java.util.NoSuchElementException
    */
-  override def head: A = throw new NoSuchElementException("Set has no elements")
+  protected def elem: A = throw new NoSuchElementException("elem of empty set")
```

Comment:
could you mention in the commit message that this commit also fixes SI-7445 for ListSet?